node/src/types/block/past_finality_signatures.rs

```diff
@@ -0,0 +1,200 @@
+#[derive(
+    Clone,
+    DataSize,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+    Debug,
+    Default,
+)]
+pub struct PastFinalitySignatures(Vec<u8>);
+
+impl PastFinalitySignatures {
+    #[allow(dead_code)] //TODO remove it when the code is used in a next ticket
+    pub fn from_validator_set<'a>(
+        public_keys: &BTreeSet<PublicKey>,
+        all_validators: impl IntoIterator<Item = &'a PublicKey>,
+    ) -> Self {
+        let bits = all_validators
+            .into_iter()
+            .map(|key| public_keys.contains(key))
+            .chunks(8)
+            .into_iter()
+            .map(|bits| {
+                bits.enumerate()
+                    .map(|(i, bit)| u8::from(bit) << (7 - i))
+                    .fold(0, |acc, u| acc | u)
+            })
+            .collect();
+
+        PastFinalitySignatures(bits)
+    }
+
+    #[allow(dead_code)] //TODO remove it when the code is used in a next ticket
+    pub fn into_validator_set(
+        self,
+        all_validators: impl IntoIterator<Item = PublicKey>,
+    ) -> BTreeSet<PublicKey> {
+        fn active(bits: u8) -> [bool; 8] {
+            [
+                0 != bits & 0b1000_0000,
+                0 != bits & 0b0100_0000,
+                0 != bits & 0b0010_0000,
+                0 != bits & 0b0001_0000,
+                0 != bits & 0b0000_1000,
+                0 != bits & 0b0000_0100,
+                0 != bits & 0b0000_0010,
+                0 != bits & 0b0000_0001,
+            ]
+        }
+
+        self.0
+            .into_iter()
+            .zip(&all_validators.into_iter().chunks(8))
+            .flat_map(|(u, validators_chunk)| {
+                IntoIterator::into_iter(active(u))
+                    .zip(validators_chunk)
+                    .filter_map(|(on, validator)| on.then_some(validator))
+            })
+            .collect()
+    }
+}
+
+impl ToBytes for PastFinalitySignatures {
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut buffer = bytesrepr::allocate_buffer(self)?;
+        let length_32: u32 = self
+            .0
+            .len()
+            .try_into()
+            .map_err(|_| bytesrepr::Error::NotRepresentable)?;
+        buffer.append(&mut length_32.to_bytes()?);
+        buffer.extend(self.0.iter().copied());
+        Ok(buffer)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.0.serialized_length()
+    }
+}
+
+impl FromBytes for PastFinalitySignatures {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (count, bytes) = u32::from_bytes(bytes)?;
+        let count = count as usize;
+        let vector = Vec::from(
+            bytes
+                .get(0..count)
+                .ok_or(bytesrepr::Error::EarlyEndOfStream)?,
+        );
+        Ok((PastFinalitySignatures(vector), &bytes[count..]))
```

Comment:
Similar here, maybe we could reuse `Vec::<u8>::from_bytes()`?

Reply:
Done